### DISCUSSION TITLE
All pages execute API on first launch and store in atom

### DISCUSSION THREAD

**Discussion by atlansien** [pmndrs/jotai#2299] [Q&A]
Nextjs14 and jotai2.6.

I am trying to access a specific GET API on first launch from all page and store the retrieved values in atom for use in each page or component.

I tried running the api in _app.tsx and storing it, but unfortunately it was not reflected in the component.

```typescript
export default function App({ Component, pageProps }: AppProps) {
  const [data, setData] = useAtom(dataAtom);
  useEffect(() => {
    const fetchData = async () => {
      try {
        const res = await fetch('http://localhost:3000/api/data');
        const data = await res.json();
        await setData(data);
      } catch (error) {
        console.error('Error during fetch:', error);
      }
    };
  }, []);

  return (
    <>
       ...
      <Provider>
          <Component {...pageProps} />
      </Provider>
    </>
  );
}

```

It seems not cool to write useEffect on every page. Do you have any suggestions?

**Comment by dai-shi** [comment]
https://jotai.org/docs/guides/nextjs

It's better to use `useHydrateAtom` for the initial render. For the second render on the client, it doesn't work and `useEffect` is only a viable solution.

**Comment by atlansien** [comment]
@dai-shi 

Thank you, I have been trying to find a way to get the information.
I have tried and failed.

```
export default function App({ Component, pageProps }: AppProps) {
  const [data, setData] = useAtom(dataAtom);
    const fetchData = async () => {
      try {
        const res = await fetch('http://localhost:3000/api/data');
        const data = await res.json();
        await setData(data);
      } catch (error) {
        console.error('Error during fetch:', error);
      }
    };

  useHydrateAtoms([[data, fetchData()]]);

  return (
    <>
       ...
      <Provider>
          <Component {...pageProps} />
      </Provider>
    </>
  );
}
```

Now trying to get it through common components. It's not great looking, but it's generally working.

**Reply by dai-shi** [reply]
useHydrateAtoms must be used within Provider.
(I think there's a recipe or something to create `HydrateAtoms` component.)

**Reply by kalijonn** [reply]
It's here: https://jotai.org/docs/guides/initialize-atom-on-render

**Comment by atlansien** [comment]
@dai-shi @kalijonn 
Thanks for the advice.
Hydration was tempting, but after trying different options, I am now using async atom.

```ts
export const asyncData = atom(async () => {
  const result = await axios.get('http://hogehoge.com/api/data');
  return result.data;
});
```

This has generally worked well.
Just trying to find a new way to declare and fetch arbitrary timing.
(For example, I want to fetch new data from the DB after POST. I want to fetch all data, so I don't change the API route or parameters.)

**Reply by dai-shi** [reply]
Does this help?
https://github.com/pmndrs/jotai/blob/1198f35b51e6b5ffff76b54d6e8a7126d6b252f0/docs/recipes/atom-with-refresh.mdx#L1-L11

**Reply by atlansien** [reply]
@dai-shi 
This is a very nice recipe. You have solved my problem.

手厚いサポート大変助かりました、ありがとうございます。

**Comment by himself65** [comment]
what is that Provider for?
